src/convopt.hpp

```diff
@@ -10,25 +10,26 @@
 
 namespace opt {
 
-// TODO(enrlov): F must be invocable with T
-template <std::size_t N, class T, class F>
-constexpr auto gradient(point<T, N> p, F cost) -> distance_t<point<T, N>>
-{
-    // TODO(olee): Should point also be a range?
-    // TODO(olee): Should point also be sized?
-    // TODO(olee): Should point be rebindable?
+// TODO(enrlov): F must be invocable with P and opt::point<dual<scalar_t<P>>>
+// and return scalar_t<...> of the input
 
-    // using T = scalar_t<distance_t<decltype(p)>>;
+template <Point P, class F>
+constexpr auto gradient(const P& p, F cost) -> distance_t<P>
+{
+    constexpr auto N = std::tuple_size_v<P>;
+    using T = scalar_t<P>;
 
     const auto d = [&p]() {
```

Comment:
Maybe this could be extracted somewhere, so a point<T> can be "promoted" to point<dual<T>>

Reply:
We might need `point` or `Point` to be have a rebindable scalar type?

Reply:
Isn't it the same T a point is build upon?